src/providers/dart_rename_provider.ts

```diff
@@ -0,0 +1,92 @@
+"use strict";
+
+import { window, workspace, RenameProvider, WorkspaceEdit, TextDocument, Position, CancellationToken, Uri, TextEdit, Range } from "vscode";
+import { Analyzer } from "../analysis/analyzer";
+import * as as from "../analysis/analysis_server_types";
+
+export class DartRenameProvider implements RenameProvider {	
+	private analyzer: Analyzer;
+	constructor(analyzer: Analyzer) {
+		this.analyzer = analyzer;
+	}
+	
+	provideRenameEdits(document: TextDocument, position: Position, newName: string, token: CancellationToken): Thenable<WorkspaceEdit> {		
+		return workspace.saveAll(false).then(() => {
+			return this.doRename(document, position, newName, token);
+		});
+	}
+
+	private doRename(document: TextDocument, position: Position, newName: string, token: CancellationToken): Thenable<WorkspaceEdit> {
+        return new Promise<WorkspaceEdit>((resolve, reject) => {
+			this.analyzer.editGetRefactoring({
+				kind: "RENAME",
+				file: document.fileName,
+				offset: document.offsetAt(position),
+				length: position.character,
+				validateOnly: false,
+				options: {
+					newName: newName
+				}			
+			}).then(resp => {				
+				let workspaceEdit = new WorkspaceEdit();
+
+				let hasError = this.handleProblem(
+					resp.initialProblems
+						.concat(resp.optionsProblems)
+						.concat(resp.finalProblems)
+				);
+
+				if (hasError) {
+					reject("Rename cancelled.");
```

Comment:
Weird, the docs do say:

> The lack of a result can be signaled by returning undefined or null.

What happens if you pass an empty array instead?

I'll try and have a play around with this later and see if I can find a way to make this better.